src/langchain_google_spanner/graph_qa.py

```diff
@@ -0,0 +1,396 @@
+            raise ValueError(
+                "Specifying gql_fix_prompt and gql_fix_llm_kwargs together is"
+                " not allowed. Please pass prompt via gql_fix_llm_kwargs."
+            )
+        if qa_prompt and qa_llm_kwargs:
+            raise ValueError(
+                "Specifying qa_prompt and qa_llm_kwargs together is"
+                " not allowed. Please pass prompt via qa_llm_kwargs."
+            )
+
+        use_qa_llm_kwargs = qa_llm_kwargs if qa_llm_kwargs is not None else {}
+        use_gql_llm_kwargs = gql_llm_kwargs if gql_llm_kwargs is not None else {}
+        use_gql_verify_llm_kwargs = (
+            gql_verify_llm_kwargs if gql_verify_llm_kwargs is not None else {}
+        )
+        use_gql_fix_llm_kwargs = (
+            gql_fix_llm_kwargs if gql_fix_llm_kwargs is not None else {}
+        )
+
+        if "prompt" not in use_qa_llm_kwargs:
+            use_qa_llm_kwargs["prompt"] = (
+                qa_prompt if qa_prompt is not None else SPANNERGRAPH_QA_PROMPT
+            )
+        if "prompt" not in use_gql_llm_kwargs:
+            use_gql_llm_kwargs["prompt"] = (
+                gql_prompt if gql_prompt is not None else GQL_GENERATION_PROMPT
+            )
+        if "prompt" not in use_gql_verify_llm_kwargs:
+            use_gql_verify_llm_kwargs["prompt"] = (
+                gql_verify_prompt
+                if gql_verify_prompt is not None
+                else GQL_VERIFY_PROMPT
+            )
+        if "prompt" not in use_gql_fix_llm_kwargs:
+            use_gql_fix_llm_kwargs["prompt"] = (
+                gql_fix_prompt if gql_fix_prompt is not None else GQL_FIX_PROMPT
+            )
+
+        gql_generation_chain = use_gql_llm_kwargs["prompt"] | llm | StrOutputParser()
+        gql_fix_chain = use_gql_fix_llm_kwargs["prompt"] | llm | StrOutputParser()
+        gql_verify_chain = (
+            use_gql_verify_llm_kwargs["prompt"] | llm | verify_gql_output_parser
+        )
+        qa_chain = use_qa_llm_kwargs["prompt"] | llm | StrOutputParser()
+
+        return cls(
+            gql_generation_chain=gql_generation_chain,
+            gql_fix_chain=gql_fix_chain,
+            gql_verify_chain=gql_verify_chain,
+            qa_chain=qa_chain,
+            **kwargs,
+        )
+
+    def execute_query(
+        self, _run_manager: CallbackManagerForChainRun, gql_query: str
+    ) -> List[Any]:
+        try:
+            _run_manager.on_text("Executing gql:", end="\n", verbose=self.verbose)
+            _run_manager.on_text(
+                gql_query, color="green", end="\n", verbose=self.verbose
+            )
+            return self.graph.query(gql_query)[: self.top_k]
+        except Exception as e:
+            raise ValueError(str(e))
+
+    def execute_with_retry(
+        self,
+        _run_manager: CallbackManagerForChainRun,
+        intermediate_steps: List,
+        question: str,
+        gql_query: str,
+    ) -> tuple[str, List[Any]]:
+        try:
```

Comment:
There seems to be some redudant exception handling logic. How about something like?

```
retries = 0

while retries < self.max_gql_fix_retries:
    try:
       return execute_query(gql_query)
   except:
      log_invalid_query()
       gql_query = fix_query()
    finally:
       retries += 1

raise ValueError(...)
```

Reply:
Good call. Made the simplification.